core/utils/thread_pool.hpp

```diff
@@ -14,40 +15,184 @@
 #include "utils/non_copyable.hpp"
 
 namespace kagome {
-  // TODO(turuslan): use `ThreadPool` in `RpcThreadPool` and `RpcContext`
+
+  class ThreadHandler final {
+    enum struct State : uint32_t { kStopped = 0, kStarted };
+
+   public:
+    ThreadHandler(ThreadHandler &&) = delete;
+    ThreadHandler(ThreadHandler const &) = delete;
+
+    ThreadHandler &operator=(ThreadHandler &&) = delete;
+    ThreadHandler &operator=(ThreadHandler const &) = delete;
+
+    ThreadHandler()
+        : execution_state_{State::kStopped},
+          ioc_{std::make_shared<boost::asio::io_context>()},
+          work_guard_{ioc_->get_executor()} {}
+
+    explicit ThreadHandler(std::shared_ptr<boost::asio::io_context> io_context)
+        : execution_state_{State::kStopped},
+          ioc_{std::move(io_context)},
+          work_guard_{} {}
+
+    ~ThreadHandler() {
+      if (work_guard_) {
+        ioc_->stop();
+      }
+    }
+
+    void start() {
+      execution_state_.store(State::kStarted, std::memory_order_release);
+    }
+
+    void stop() {
+      execution_state_.store(State::kStopped, std::memory_order_release);
+    }
+
+    template <typename F>
+    void execute(F &&func) {
+      BOOST_ASSERT(ioc_);
+      if (State::kStarted == execution_state_.load(std::memory_order_acquire)) {
+        ioc_->post(std::forward<F>(func));
+      }
+    }
+
+    template <typename F, typename... Args>
+    auto reinvoke(F &&func, Args &&...args)
+        -> std::optional<std::tuple<Args...>> {
+      if (!isInCurrentThread()) {
+        execute([func(std::forward<F>(func)),
+                 tup{std::tuple<std::decay_t<Args>...>{
+                     std::forward<Args>(args)...}}]() mutable {
+          std::apply([&](auto &&...a) mutable { func(std::move(a)...); },
+                     std::move(tup));
+        });
+        return std::nullopt;
+      }
+      return {std::make_tuple(std::forward<Args>(args)...)};
+    }
+
+    bool isInCurrentThread() const {
+      BOOST_ASSERT(ioc_);
+      return ioc_->get_executor().running_in_this_thread();
+    }
```

Comment:
Why not just async `io_context.post(F)` (stopped io_context won't execute F, and means app shutdown)?

Why not single free function `run_in_thread(std::io_context &, F)` otherwise?